crates/steel/src/history/eip2935/mod.rs

```diff
@@ -0,0 +1,223 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types related to commitments to a historical state relying on EIP-2935.
+
+use alloy_primitives::{Sealed, B256, U256};
+use history_storage::HistoryStorageContract;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    history::state::SingleContractState, BlockHeaderCommit, Commitment, CommitmentVersion,
+    ComposeInput, EvmBlockHeader, EvmFactory,
+};
+
+mod history_storage;
+
+/// Input recursively committing to multiple execution block hashes.
+pub type HistoryInput<F> = ComposeInput<F, HistoryCommit<<F as EvmFactory>::Header>>;
+
+/// Commitment that an execution block is an ancestor of a specific other execution block.
+///
+/// This struct encapsulates the necessary data to prove that a given execution block is part of the
+/// canonical chain.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct HistoryCommit<H> {
+    /// Iterative commits for verifying the execution block as an ancestor of some other block.
+    state_commits: Vec<StateCommit<H>>,
+}
+
+/// Represents a commitment of a history storage contract state to the corresponding block hash.
+#[derive(Clone, Serialize, Deserialize)]
+struct StateCommit<H> {
+    /// State for verifying the previous execution block hash.
+    state: SingleContractState,
+    /// Header belonging to the state.
+    header: H,
+}
+
+impl<H: EvmBlockHeader> BlockHeaderCommit<H> for HistoryCommit<H> {
+    /// Generates a commitment that proves the given block header is part of the chain history.
+    ///
+    /// Panics if the provided [HistoryCommit] data is invalid or inconsistent.
+    fn commit(mut self, header: &Sealed<H>, config_id: B256) -> Commitment {
+        let mut header = header.as_sealed_ref();
+
+        // starting from header, "walk forward" along state_commits to reach a later execution hash
+        for state_commit in &mut self.state_commits {
+            let state_header = state_commit.header.seal_ref_slow();
+
+            // verify that the block to query is in the allowed history window
+            assert!(
+                header.number() < state_header.number()
+                    && state_header.number() - header.number()
+                        <= history_storage::HISTORY_SERVE_WINDOW.to(),
+                "Block outside of EIP-2935 history range"
+            );
+
+            // verify that the state is valid with respect to the commitment header
+            assert_eq!(
+                &state_commit.state.root(),
+                state_header.state_root(),
+                "State root mismatch"
+            );
+
+            let block_number = U256::from(header.number());
+            let execution_hash = HistoryStorageContract::new(&mut state_commit.state)
+                .and_then(|mut c| c.get_unchecked(block_number))
+                .expect("History storage contract failed");
```

Comment:
This does not verify that the current specification supports EIP-2935 because the current `Commitment` traits do not support it. The `HistoryStorageContract` implementation verifies that the deployed contract code is correct. However, there is a small potential security risk on chains without EIP-2935 support: The contract could have been deployed with malicious state, or the stored hashes could belong to something different on that particular chain.